# Lesson 4 Assignment
### Jessica Kunkel

In [1]:
#!/usr/bin/env python
""" Import display functions and create a function to print outputs to 
Markdown """

from IPython.display import display, Markdown
def printmd(string):
    display(Markdown(string))

In [2]:
""" Import NumPy, pandas, and dataframe """
import numpy as np
import pandas as pd
from pandas import DataFrame

## Question 1: Data Cleaning, Processing, and Manipulating


First, I changed the working directory of the Jupyter notebook to the directory where I saved this week's data file. 

In [3]:
""" Import the os module and set the working directory to where the 
mtcars.csv file is saved """

import os
path = r"C:\Users\jkunk\OneDrive\Documents\_SWENG Masters\9 DAAN 862 - Analytics Programming Python"
os.chdir(path)

### 1. Explore the datasets.

I then used the pandas `read_csv` function to import the Assignment4_data file. I chose to display the first 5 rows of the data using the `head` function.

In [4]:
""" use read_csv to import Assignment4_data.csv, show the first 5 rows """

assn4 = pd.read_csv("Assignment4_data.csv")
assn4.head()

one   two  three  four  five variable
0 -92.0 -76.0  -33.0   3.0 -13.0       B2
1 -21.0  76.0   38.0  -6.0  80.0       B1
2  -2.0 -47.0  -34.0 -86.0 -66.0       A1
3 -76.0  43.0    7.0 -40.0 -42.0       A1
4  44.0  37.0   -7.0 -14.0  30.0       A1

In order to gain more information on what kind of data was in the assn4 dataset, I used the `info` function. 

In [5]:
""" get information on the data included in the assn4 dataset """

assn4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   one       195 non-null    float64
 1   two       197 non-null    float64
 2   three     199 non-null    float64
 3   four      194 non-null    float64
 4   five      196 non-null    float64
 5   variable  200 non-null    object 
dtypes: float64(5), object(1)
memory usage: 9.5+ KB


The output of the `info` function showed that the majority of elements in the dataset are non-null, but all columns except "variable" have at least one null value. In order to explore the data a little further, I chose to look into 50 rows in the dataset. 

In [6]:
""" display rows 100-149 of the assn4 dataframe """

assn4[100:150]

one   two  three  four   five variable
100   20.0 -53.0   13.0   4.0   85.0       A1
101  -74.0 -18.0   44.0 -49.0  -13.0       B2
102  -90.0  96.0   67.0  -2.0   28.0       B2
103  -99.0  32.0  -42.0  69.0  -94.0       B1
104  -63.0 -16.0  -63.0 -59.0   65.0       B1
105   95.0  75.0   60.0 -68.0   50.0       A1
106  -72.0 -44.0  -14.0  66.0   29.0       A2
107   77.0  -9.0   99.0   9.0 -821.0       A1
108   84.0 -86.0  -95.0  -4.0   84.0       B1
109  -73.0  -6.0  -86.0  70.0   -8.0       A1
110   13.0 -13.0   24.0 -71.0  -19.0       A2
111   78.0  97.0  -93.0 -96.0   94.0       A2
112  -81.0 -14.0  -18.0  -2.0   80.0       B2
113  -17.0 -24.0   39.0 -63.0  -21.0       B1
114   23.0  37.0  -89.0 -70.0   97.0       A1
115   93.0 -49.0  776.0  83.0   15.0       B2
116 -363.0 -64.0  -61.0 -18.0   31.0       A2
117  -67.0  65.0  -59.0 -47.0  -94.0       A1
118   12.0 -86.0   26.0 -82.0  -82.0       A2
119   27.0  -6.0   47.0 -16.0  -85.0       A1
120  -23.0 -75.0  -53.0  65.0  -43.0       B2
121  -74.0  10.0   87.0 -47.0   74.0       B1
122  -53.0 -14.0   13.0 -55.0   88.0       B2
123   72.0  36.0   -7.0  35.0   81.0       A1
124  -64.0  92.0   62.0 -96.0   40.0       A1
125   68.0 -66.0   -7.0  23.0  -85.0       B2
126   40.0 -15.0   51.0 -33.0    NaN       B2
127  -40.0 -21.0   37.0 -37.0   83.0       B1
128  -28.0  44.0   27.0  54.0  -73.0       B1
129  -12.0  62.0   51.0 -42.0   40.0       A2
130   72.0  82.0   35.0   0.0   79.0       A1
131  -18.0  69.0  -71.0 -25.0    NaN       B2
132   42.0 -91.0  -22.0  50.0   55.0       A1
133  -45.0  30.0   11.0 -80.0   27.0       B1
134   74.0 -42.0  -93.0 -40.0   28.0       B1
135   93.0  94.0  -80.0  92.0  -22.0       B1
136  -33.0  36.0   -1.0  40.0   23.0       A1
137   -5.0  48.0  -63.0  68.0  -18.0       B2
138  -73.0  39.0  -29.0 -64.0  -53.0       B1
139   90.0   2.0   47.0   NaN   63.0       A1
140   29.0  -5.0  -54.0 -48.0   15.0       A1
141   -8.0  41.0   50.0 -17.0  -25.0       A2
142  -89.0 -69.0   66.0 -51.0   44.0       A1
143  -56.0  23.0  -65.0 -42.0   48.0       A2
144   65.0 -37.0  -97.0 -36.0   98.0       A2
145   28.0   4.0   88.0  44.0   99.0       B1
146  -12.0 -39.0  -11.0 -58.0  -34.0       A2
147  -41.0  24.0   50.0   NaN  -47.0       B1
148  -32.0  61.0   -5.0 -15.0  -31.0       B2
149    0.0  16.0   55.0 -64.0  -34.0       A1

### 2. Find and handle missing values in the data. (It is your choice how you handle the missing data.)

While the majority of values in each row are non-zero, there were two elements whose value was 0.0 in the dataset, so using the `fillna` function to replace NaN values with 0 could potentially confuse the existing data. Therefore, I chose to use the `dropna` method to throw out rows with null values. I think, in terms of scientific experimentation, if you're observing 5 variables and have an invalid reading on one of the variables for a test run, that run would generally be discarded and not considered in the evaluation of the data.

I decided not to modify the original dataset, and instead save the cleaned dataset to a new dataframe `assn4_nonull`, which will be used moving forward.

In [7]:
""" use dropna to remove rows containing null values from the dataset, 
saving the result to a new dataframe assn4_nonull. Display the new 
dataframe. """

assn4_nonull = assn4.dropna() 
assn4_nonull

one   two  three  four  five variable
0   -92.0 -76.0  -33.0   3.0 -13.0       B2
1   -21.0  76.0   38.0  -6.0  80.0       B1
2    -2.0 -47.0  -34.0 -86.0 -66.0       A1
3   -76.0  43.0    7.0 -40.0 -42.0       A1
4    44.0  37.0   -7.0 -14.0  30.0       A1
..    ...   ...    ...   ...   ...      ...
195  63.0   3.0  -30.0 -24.0 -59.0       A1
196  97.0 -48.0  -61.0 -25.0 -21.0       B1
197 -93.0 -75.0  -18.0 -67.0 -58.0       B1
198  54.0 -66.0  -80.0  92.0  62.0       A1
199  82.0  53.0  -77.0  79.0  97.0       B2

[181 rows x 6 columns]

In total, 9 rows were discarded due to null values. 

### 3. Explore the variable column and convert the "variable” column to dummy variables and join the dummies to the data.

From the previous exploration of the data, it looks like the "variable" column includes values [A1, B1, A2, B2]. In order to convert the "variable" column to dummy variables, I used the `get_dummies` method. In order to align with this week's reading and the textbook, I chose to keep the data type that the `get_dummies` function returned as binary 0 or 1, rather than Boolean True or False. I saved these dummy variables to `dummies`

In [8]:
""" use get_dummies to get dummy values for the variables column. Save 
the result to a new dataframe called dummies and display it. """

dummies = pd.get_dummies(assn4_nonull['variable'], dtype=int)
dummies

A1  A2  B1  B2
0     0   0   0   1
1     0   0   1   0
2     1   0   0   0
3     1   0   0   0
4     1   0   0   0
..   ..  ..  ..  ..
195   1   0   0   0
196   0   0   1   0
197   0   0   1   0
198   1   0   0   0
199   0   0   0   1

[181 rows x 4 columns]

I joined the dummy variables to the data using the `join` function, saving the result to `assn4_nonull_dummy`. For clarity in the dataset, I chose to disinclude the "variables" column in the new dataframe (selecting columns 'one', 'two', 'three', 'four', and 'five' to remain in the joined dataframe), so that the dummy columns were more consistent to view.

In [9]:
""" use the join function to join the dummies dataframe to the 
assn4_nonull dataframe, saving the result to a new dataframe 
assn4_nonull_dummy. The columns 'one', 'two', 'three', 'four', 
and 'five' are included in the join operation. Display the new
dataframe. """

assn4_nonull_dummy = assn4_nonull[['one', 'two', 'three', 'four', 'five']].join(dummies)
assn4_nonull_dummy

one   two  three  four  five  A1  A2  B1  B2
0   -92.0 -76.0  -33.0   3.0 -13.0   0   0   0   1
1   -21.0  76.0   38.0  -6.0  80.0   0   0   1   0
2    -2.0 -47.0  -34.0 -86.0 -66.0   1   0   0   0
3   -76.0  43.0    7.0 -40.0 -42.0   1   0   0   0
4    44.0  37.0   -7.0 -14.0  30.0   1   0   0   0
..    ...   ...    ...   ...   ...  ..  ..  ..  ..
195  63.0   3.0  -30.0 -24.0 -59.0   1   0   0   0
196  97.0 -48.0  -61.0 -25.0 -21.0   0   0   1   0
197 -93.0 -75.0  -18.0 -67.0 -58.0   0   0   1   0
198  54.0 -66.0  -80.0  92.0  62.0   1   0   0   0
199  82.0  53.0  -77.0  79.0  97.0   0   0   0   1

[181 rows x 9 columns]

### 4. Convert the "one” column into 3 bins.

In order to convert the "one" column into 3 bins, I used the `cut` function. Since the edge values for each bins were not specified, I chose to specify the number of bins in the `cut` function. I also chose to define `precision=1` to be consistent with the values in the dataset.

In [10]:
""" use the cut function to convert the 'one' column into 3 bins, 
specifying the number of bins and precision=1 for consistency. """

pd.cut(assn4_nonull_dummy['one'], 3, precision=1)

0      (-209.7, -56.3]
1        (-56.3, 97.0]
2        (-56.3, 97.0]
3      (-209.7, -56.3]
4        (-56.3, 97.0]
            ...       
195      (-56.3, 97.0]
196      (-56.3, 97.0]
197    (-209.7, -56.3]
198      (-56.3, 97.0]
199      (-56.3, 97.0]
Name: one, Length: 181, dtype: category
Categories (3, interval[float64, right]): [(-363.5, -209.7] < (-209.7, -56.3] < (-56.3, 97.0]]

The bins specified by `cut` are as follows:

    Bin 1: (-363.5, -209.7] 
    Bin 2: (-209.7, -56.3]     
    Bin 3: (-56.3, 97.0]

## Question 2: String Manipulation

In [11]:
""" Load the assigned string into the Jupyter notebook """

s = "I am happy to join with you today in what will go down in history as the greatest demonstration for freedom in the history of our nation. Five score years ago, a great American, in whose symbolic shadow we stand today, signed the Emancipation Proclamation. This momentous decree came as a great beacon light of hope to millions of Negro slaves who had been seared in the flames of withering injustice. It came as a joyous daybreak to end the long night of their captivity. But one hundred years later, the Negro still is not free. One hundred years later, the life of the Negro is still sadly crippled by the manacles of segregation and the chains of discrimination. One hundred years later, the Negro lives on a lonely island of poverty in the midst of a vast ocean of material prosperity. One hundred years later, the Negro is still languishing in the corners of American society and finds himself an exile in his own land. So we have come here today to dramatize a shameful condition."

### 1. Find out how many unique words in s.

To start, I chose to convert all the characters to lowercase. Looking through the readings for this week, I did not find any one function that would quickly solve this task, so I decided that splitting the string into a string array would likely be a logical step in my solution. None of the questions were concerned with punctuation or the number of sentences within the string, so I also decided to remove all the punctuation from the string. 

In [12]:
""" prepare the string array:  remove punctuation marks, convert 
all characters to lowercase, and split the string into an array."""

# import string in order to use the predefined string.punctuation
import string

# use comprehension in the join function to remove punctuation marks
s_clean = ''.join(char for char in s if char not in string.punctuation)
# convert to lowercase
s_lower = s_clean.lower()
# create the string array by splitting using a space ' ' as the delimiter
s_arr = s_lower.split(' ')

In order to find the number of unique words in the string, I decided to use a for-loop. I created an empty string `unique` and for each word in the string array, if the word was not already in `unique`, appended the word to the array. or comparative purposes, I chose to print, in markdown, the number of words in the initial string array versus the number of unique words in the speech.

In [13]:
""" find the number of unique words in the speech, using the 
string array s_arr """

unique = []
for s in s_arr:
    if s not in unique:
        unique.append(s)

len(unique)
printmd(f"The number of words in the string is {len(s_arr)}. The number of unique words in the string is {len(unique)}.")

The number of words in the string is 177. The number of unique words in the string is 107.

### 2. Which word appears the most?

To determine which word appears the most, I decided to make a dictionary of unique words and the number of times they appear in the speech. I used a dictionary comprehension, where the key is a word in the `unique` list, and the value is the count of times the word appears in the initial `s_arr`. For demonstrative purposes, I chose to print the dictionary of unique words and the number of times they appear in the speech.

In [14]:
""" create a dictionary of unique words and the number of times 
they appear in the speech s_arr """

unique_dict = {word: s_arr.count(word) for word in unique}
unique_dict

{'i': 1,
 'am': 1,
 'happy': 1,
 'to': 4,
 'join': 1,
 'with': 1,
 'you': 1,
 'today': 3,
 'in': 8,
 'what': 1,
 'will': 1,
 'go': 1,
 'down': 1,
 'history': 2,
 'as': 3,
 'the': 14,
 'greatest': 1,
 'demonstration': 1,
 'for': 1,
 'freedom': 1,
 'of': 12,
 'our': 1,
 'nation': 1,
 'five': 1,
 'score': 1,
 'years': 5,
 'ago': 1,
 'a': 6,
 'great': 2,
 'american': 2,
 'whose': 1,
 'symbolic': 1,
 'shadow': 1,
 'we': 2,
 'stand': 1,
 'signed': 1,
 'emancipation': 1,
 'proclamation': 1,
 'this': 1,
 'momentous': 1,
 'decree': 1,
 'came': 2,
 'beacon': 1,
 'light': 1,
 'hope': 1,
 'millions': 1,
 'negro': 5,
 'slaves': 1,
 'who': 1,
 'had': 1,
 'been': 1,
 'seared': 1,
 'flames': 1,
 'withering': 1,
 'injustice': 1,
 'it': 1,
 'joyous': 1,
 'daybreak': 1,
 'end': 1,
 'long': 1,
 'night': 1,
 'their': 1,
 'captivity': 1,
 'but': 1,
 'one': 4,
 'hundred': 4,
 'later': 4,
 'still': 3,
 'is': 3,
 'not': 1,
 'free': 1,
 'life': 1,
 'sadly': 1,
 'crippled': 1,
 'by': 1,
 'manacles': 1,
 'segrega

I then used dictionary functions to determine the word that appears the most in the speech and the number of occurrences.

In [15]:
""" find the maximum value in the dictionary and the 
corresponding key """

# find the maximum value in the dict
max_count = max(unique_dict.values())
# find the word with the maximum value in the dict
max_word = max(unique_dict, key=unique_dict.get)

# print, in markdown, the word with the most uses and its number of occurrences
printmd(f"The word that appears the most in the speech is \"{max_word}\" with {max_count} occurrences.")

The word that appears the most in the speech is "the" with 14 occurrences.

### 3. How many words start with ‘t’?

In order to count the number of words that start with the letter 't', I chose to use a for-loop. I initialized a variable `count` to count the number of words, then looped through the list of unique words in the speech, incrementing the count if `word.startswith("t")` returned true. I then printed, in markdown, the number of words in the speech that start with the letter 't'.

In [16]:
""" use a simple for-loop and the startswith method to 
find the number of words in the speech that start with the 
letter 't' """

# initialize count
count = 0

# use startswith method to test if each word in the unique word list starts with the letter 't'
for word in unique:
    if word.startswith("t"):
        # if true, increment count by 1
        count += 1

# print, in markdown, the number of words that start with the letter 't'
printmd(f"The number of words in the speech that start with the letter \'t\' is {count}.")

The number of words in the speech that start with the letter 't' is 5.